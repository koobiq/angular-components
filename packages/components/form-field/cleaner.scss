@use 'sass:meta';
@use 'sass:map';

@use '../core/styles/common/tokens' as *;

.kbq-cleaner {
    display: flex;
    align-items: center;
    justify-content: center;

    width: kbq-css-variable(form-field-size-icon-button-size);
    margin-right: kbq-css-variable(form-field-size-icon-button-margin-right);
    height: 100%;

    cursor: pointer;
}
