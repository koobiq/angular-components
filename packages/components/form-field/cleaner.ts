import { Component, ViewEncapsulation } from '@angular/core';

@Component({
    selector: 'kbq-cleaner',
    exportAs: 'kbqCleaner',
    template: `
        <i [autoColor]="true" kbq-icon-button="kbq-xmark-circle_16"></i>
    `,
    styleUrls: ['cleaner.scss'],
    host: {
        class: 'kbq-cleaner'
    },
    encapsulation: ViewEncapsulation.None
})
export class KbqCleaner {}
