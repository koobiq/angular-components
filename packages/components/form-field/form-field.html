<div
    #connectionContainer
    class="kbq-form-field__container"
    (click)="onContainerClick($event)"
>
    @if (hasPrefix) {
        <div class="kbq-form-field__prefix">
            <ng-content select="[kbqPrefix]" />
        </div>
    }

    <div class="kbq-form-field__infix">
        <ng-content />
    </div>

    @if (hasSuffix) {
        <div class="kbq-form-field__suffix">
            <ng-content select="[kbqSuffix]" />
        </div>
    }

    @if (canShowCleaner && !hasSuffix) {
        <div
            class="kbq-form-field__cleaner"
            (click)="clearValue($event)"
        >
            <ng-content select="kbq-cleaner" />
        </div>
    }

    <ng-content select="kbq-password-toggle" />

    @if (canShowStepper) {
        <ng-content select="kbq-stepper" />
    }
</div>

@if (hasHint || passwordHints.length) {
    <div class="kbq-form-field__hint">
        <ng-content select="kbq-hint, kbq-password-hint" />
    </div>
}
