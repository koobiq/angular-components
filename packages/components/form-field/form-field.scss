/* stylelint-disable scss/operator-no-newline-after */

@use 'sass:meta';
@use 'sass:map';

@use '../core/forms/forms';

@use '../core/styles/common/tokens' as *;

@use './form-field-theme' as *;

.kbq-form-field {
    position: relative;
    display: inline-block;
    width: 100%;

    .kbq-form-field__container {
        box-sizing: border-box;
    }

    .kbq-form-field:not(.kbq-form-field-type-textarea) {
        height: var(--kbq-form-field-size-height);
    }

    &,
    .kbq-input,
    .kbq-textarea {
        border-radius: var(--kbq-form-field-size-border-radius);
    }

    &:hover {
        z-index: 1;
    }

    &.kbq-focused {
        z-index: 2;
    }

    &.kbq-form-field_without-borders .kbq-form-field__container {
        border-color: transparent !important;
        box-shadow: none !important;
    }

    &.kbq-form-field_has-prefix {
        .kbq-input {
            padding-left: kbq-sum-series-css-variables(
                [form-field-size-icon-size,
                form-field-size-icon-margin-left,
                form-field-size-icon-margin-right]
            );
        }
    }

    &.kbq-form-field_has-password-toggle,
    &.kbq-form-field_has-cleaner,
    &.kbq-form-field_has-stepper,
    &.kbq-form-field_has-suffix {
        .kbq-input {
            padding-right: kbq-sum-series-css-variables(
                [form-field-size-icon-button-size,
                form-field-size-icon-button-margin-left,
                form-field-size-icon-button-margin-right]
            );
        }
    }

    & + .kbq-password-hint {
        margin-top: var(--kbq-size-m);
    }
}

.kbq-form-field__container {
    position: relative;

    border-width: var(--kbq-form-field-size-border-width);
    border-style: solid;
    border-color: transparent;
    border-radius: var(--kbq-form-field-size-border-radius);
}

.kbq-form-field__prefix,
.kbq-form-field__suffix {
    position: absolute;
    top: 0;
    bottom: 0;
    width: 32px;

    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
}

.kbq-form-field__prefix {
    left: 0;
}

.kbq-form-field__suffix {
    right: 0;
}

.kbq-form-field__cleaner,
.kbq-password-toggle,
.kbq-stepper {
    position: absolute;

    top: 0;
    bottom: 0;
    right: 0;
}

.kbq-stepper {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;

    width: var(--kbq-form-field-size-icon-button-size);

    .kbq-stepper-step-up,
    .kbq-stepper-step-down {
        cursor: pointer;
        width: var(--kbq-form-field-size-icon-button-size);
        text-align: center;
    }

    .kbq-stepper-step-up {
        transform: scaleY(-1);
    }
}

.kbq-form-field__hint {
    display: flex;
    flex-direction: column;
    margin-top: var(--kbq-form-field-hint-size-margin-top);

    & .kbq-password-hint:first-child {
        margin-top: var(--kbq-size-s);
    }
}

.kbq-password-hint {
    margin-top: var(--kbq-form-field-hint-size-gap);
}

@include kbq-form-field-theme();
@include kbq-form-field-typography();
