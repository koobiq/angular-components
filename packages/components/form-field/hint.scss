@use 'sass:meta';
@use 'sass:map';

@use '../core/styles/common/tokens' as *;

@use './hint-theme' as *;

.kbq-hint {
    display: flex;

    & .kbq-icon {
        margin-top: kbq-css-variable(hint-size-normal-margin-top);
        margin-right: kbq-css-variable(hint-size-normal-content-padding);
    }

    &.kbq-hint_compact {
        & .kbq-icon {
            margin-top: kbq-css-variable(hint-size-compact-margin-top);
            margin-right: kbq-css-variable(hint-size-compact-content-padding);
        }
    }
}

@include kbq-hint-theme();
@include kbq-hint-typography();
