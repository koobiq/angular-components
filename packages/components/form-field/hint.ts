import { ChangeDetectionStrategy, Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
import { CanColorCtor, KbqComponentColors, mixinColor, toBoolean } from '@koobiq/components/core';

let nextHintUniqueId = 0;

/** @docs-private */
export class KbqHintBase {
    constructor(public elementRef: ElementRef) {}
}

/** @docs-private */
export const KbqHintMixinBase: CanColorCtor & typeof KbqHintBase = mixinColor(KbqHintBase, KbqComponentColors.Empty);

@Component({
    selector: 'kbq-hint',
    exportAs: 'kbqHint',
    template: `
        <ng-content select="[kbq-icon]"></ng-content>

        <span class="kbq-hint__text">
            <ng-content></ng-content>
        </span>
    `,
    styleUrls: ['hint.scss'],
    inputs: ['color'],
    host: {
        class: 'kbq-hint',
        '[attr.id]': 'id',
        '[class.kbq-hint_fill-text-off]': 'fillTextOff',
        '[class.kbq-hint_compact]': 'compact'
    },
    encapsulation: ViewEncapsulation.None,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class KbqHint extends KbqHintMixinBase {
    @Input() id: string = `kbq-hint-${nextHintUniqueId++}`;

    @Input() fillTextOff: boolean = false;

    @Input()
    get compact() {
        return this._compact;
    }

    set compact(value: any) {
        this._compact = toBoolean(value);
    }

    private _compact = false;

    constructor(elementRef: ElementRef) {
        super(elementRef);
    }
}
