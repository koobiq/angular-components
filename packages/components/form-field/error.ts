import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { KbqComponentColors } from '@koobiq/components/core';
import { KbqHint } from './hint';

/** Error text to be shown below the form field control. */
@Component({
    standalone: true,
    selector: 'kbq-error',
    exportAs: 'kbqError',
    templateUrl: './hint.html',
    styleUrls: [
        './hint.scss',
        './hint-tokens.scss'
    ],
    host: {
        class: 'kbq-error'
    },
    encapsulation: ViewEncapsulation.None,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class KbqError extends KbqHint {
    constructor() {
        super();

        this.color = KbqComponentColors.Error;
    }
}
