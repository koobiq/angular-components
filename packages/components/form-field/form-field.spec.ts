import { Component, DebugElement, Type } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import {
    AbstractControl,
    FormControl,
    FormGroup,
    FormGroupDirective,
    NgForm,
    ReactiveFormsModule,
    Validators
} from '@angular/forms';
import { By } from '@angular/platform-browser';
import {
    ErrorStateMatcher,
    kbqDisableLegacyValidationDirectiveProvider,
    PasswordValidators,
    ShowOnFormSubmitErrorStateMatcher
} from '@koobiq/components/core';
import { KbqInput, KbqInputModule, KbqInputPassword } from '@koobiq/components/input';
import { KbqCleaner } from './cleaner';
import { KbqError } from './error';
import { getKbqFormFieldMissingControlError, KbqFormField, kbqFormFieldDefaultOptionsProvider } from './form-field';
import { KbqFormFieldModule } from './form-field.module';
import { KbqHint } from './hint';
import { KbqLabel } from './label';
import { KbqPasswordToggle } from './password-toggle';
import { KbqPrefix } from './prefix';
import { KbqReactivePasswordHint } from './reactive-password-hint';
import { KbqSuffix } from './suffix';

const createComponent = <T>(component: Type<T>, providers: any[] = []): ComponentFixture<T> => {
    TestBed.configureTestingModule({ imports: [component], providers }).compileComponents();
    const fixture = TestBed.createComponent<T>(component);

    fixture.autoDetectChanges();

    return fixture;
};

const getErrorDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqError));
};

const getHintDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqHint));
};

const getCleanerDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqCleaner));
};

const getFormFieldDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqFormField));
};

const getLabelDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqLabel));
};

const getSuffixDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqSuffix));
};

const getPasswordToggleDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqPasswordToggle));
};

const getReactivePasswordHintDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqReactivePasswordHint));
};

const getLabelNativeElement = (debugElement: DebugElement): HTMLLabelElement => {
    return debugElement.query(By.css('label')).nativeElement;
};

const getPrefixDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqPrefix));
};

const getInputDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqInput));
};

const getInputNativeElement = (debugElement: DebugElement): HTMLInputElement => {
    return getInputDebugElement(debugElement).nativeElement;
};

const getPasswordInputNativeElement = (debugElement: DebugElement): HTMLInputElement => {
    return debugElement.query(By.directive(KbqInputPassword)).nativeElement;
};

const getSubmitButtonNativeElement = (debugElement: DebugElement): HTMLInputElement => {
    return debugElement.query(By.css('[type="submit"]')).nativeElement;
};

@Component({
    selector: 'input-form-field-with-hint-and-error',
    standalone: true,
    imports: [KbqFormFieldModule, KbqInputModule, ReactiveFormsModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <kbq-form-field>
            <input kbqInput [formControl]="control" />
            <kbq-hint id="test-hint-id">Hint</kbq-hint>
            <kbq-error id="test-error-id">Error</kbq-error>
        </kbq-form-field>
    `
})
export class InputFormFieldWithHintAndError {
    readonly control = new FormControl('', [Validators.required]);
}

@Component({
    selector: 'input-form-field-with-prefix-and-suffix',
    standalone: true,
    imports: [KbqFormFieldModule, KbqInputModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <kbq-form-field>
            <input kbqInput />
            <span kbqPrefix>Prefix</span>
            <span kbqSuffix>Suffix</span>
        </kbq-form-field>
    `
})
export class InputFormFieldWithPrefixAndSuffix {}

@Component({
    selector: 'input-form-field-with-cleaner',
    standalone: true,
    imports: [KbqFormFieldModule, KbqInputModule, ReactiveFormsModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <kbq-form-field>
            <input kbqInput [formControl]="control" />
            <kbq-cleaner />
        </kbq-form-field>
    `
})
export class InputFormFieldWithCleaner {
    readonly control = new FormControl();
}

@Component({
    selector: 'input-form-field-without-form-field-control',
    standalone: true,
    imports: [KbqFormFieldModule, ReactiveFormsModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <kbq-form-field>
            <input [formControl]="control" />
        </kbq-form-field>
    `
})
export class InputFormFieldWithoutFormFieldControl {
    readonly control = new FormControl();
}

@Component({
    selector: 'input-form-field-with-label',
    standalone: true,
    imports: [KbqFormFieldModule, ReactiveFormsModule, KbqInputModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <kbq-form-field>
            <kbq-label>Label</kbq-label>
            <input kbqInput [id]="id" />
        </kbq-form-field>
    `
})
export class InputFormFieldWithLabel {
    readonly id = 'UNIQUE_TEST_ID';
}

class CustomErrorStateMatcher implements ErrorStateMatcher {
    isErrorState(control: AbstractControl | null, _form: FormGroupDirective | NgForm | null): boolean {
        return !!control?.invalid;
    }
}

@Component({
    selector: 'input-form-field-with-custom-error-state-matcher',
    standalone: true,
    imports: [KbqFormFieldModule, KbqInputModule, ReactiveFormsModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <form [formGroup]="formGroup">
            <kbq-form-field>
                <input formControlName="email" kbqInput [errorStateMatcher]="errorStateMatcher" />
                <kbq-error id="test-error-id">Error</kbq-error>
            </kbq-form-field>
            <input type="submit" />
        </form>
    `
})
export class InputFormFieldWithCustomErrorStateMatcher {
    readonly formGroup = new FormGroup({
        email: new FormControl('', [Validators.required, Validators.email])
    });
    errorStateMatcher = new CustomErrorStateMatcher();
}

@Component({
    selector: 'input-form-field-with-border-customization',
    standalone: true,
    imports: [KbqFormFieldModule, ReactiveFormsModule, KbqInputModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <kbq-form-field [noBorders]="noBorders">
            <input kbqInput />
        </kbq-form-field>
    `
})
export class InputFormFieldWithBorderCustomization {
    noBorders: boolean = true;
}

@Component({
    selector: 'password-form-field',
    standalone: true,
    imports: [KbqFormFieldModule, ReactiveFormsModule, KbqInputModule],
    providers: [kbqDisableLegacyValidationDirectiveProvider()],
    template: `
        <kbq-form-field>
            <input kbqInputPassword [formControl]="formControl" />
            <kbq-password-toggle />
            <kbq-reactive-password-hint
                id="test-reactive-password-hint-id"
                [hasError]="formControl.hasError('minLength')"
            >
                Min length
                @let minLength = formControl.getError('minLength');
                @if (minLength) {
                    ({{ minLength.actual }}/{{ minLength.min }})
                }
            </kbq-reactive-password-hint>
        </kbq-form-field>
    `
})
export class PasswordFormField {
    readonly formControl = new FormControl('', [PasswordValidators.minLength(8)]);
}

@Component({
    selector: 'input-form-field-with-legacy-validation-directive',
    standalone: true,
    imports: [KbqFormFieldModule, KbqInputModule],
    template: `
        <kbq-form-field>
            <input kbqInput />
        </kbq-form-field>
    `
})
export class InputFormFieldWithLegacyValidationDirective {}

describe(KbqFormField.name, () => {
    it('should display KbqHint', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);

        expect(getHintDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should display KbqError', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);
        const input = getInputNativeElement(debugElement);

        input.focus();
        input.blur();
        expect(getErrorDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should hide KbqError', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);

        expect(getErrorDebugElement(debugElement)).toBeNull();
    });

    it('should display KbqPrefix', () => {
        const { debugElement } = createComponent(InputFormFieldWithPrefixAndSuffix);

        expect(getPrefixDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should display KbqSuffix', async () => {
        const { debugElement } = createComponent(InputFormFieldWithPrefixAndSuffix);

        expect(getSuffixDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should hide KbqCleaner', () => {
        const { debugElement } = createComponent(InputFormFieldWithCleaner);

        expect(getCleanerDebugElement(debugElement)).toBeNull();
    });

    it('should display KbqCleaner', () => {
        const { debugElement } = createComponent(InputFormFieldWithCleaner);
        const input = getInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));
        expect(getCleanerDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should clean field by KbqCleaner', () => {
        const { debugElement, componentInstance } = createComponent(InputFormFieldWithCleaner);
        const input = getInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));
        expect(componentInstance.control.value).toBe('koobiq');
        const cleaner = getCleanerDebugElement(debugElement);

        cleaner.nativeElement.click();
        expect(componentInstance.control.value).toBeNull();
    });

    it('should throw Error for KbqFormField without KbqFormFieldControl', () => {
        let caught = false;

        try {
            createComponent(InputFormFieldWithoutFormFieldControl);
        } catch (error) {
            expect(error.message).toBe(getKbqFormFieldMissingControlError().message);
            caught = true;
        }

        expect(caught).toBeTruthy();
    });

    it('should add ng-untouched selector for KbqFormField initially', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);

        expect(getFormFieldDebugElement(debugElement).classes['ng-untouched']).toBeTruthy();
    });

    it('should add ng-touched selector for KbqFormField after blur', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);
        const formField = getFormFieldDebugElement(debugElement);

        expect(formField.classes['ng-touched']).toBeFalsy();
        const input = getInputNativeElement(debugElement);

        input.focus();
        input.blur();
        expect(formField.classes['ng-touched']).toBeTruthy();
    });

    it('should add ng-pristine selector for KbqFormField initially', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);

        expect(getFormFieldDebugElement(debugElement).classes['ng-pristine']).toBeTruthy();
    });

    it('should add ng-dirty selector for KbqFormField after form control change', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);
        const formField = getFormFieldDebugElement(debugElement);

        expect(formField.classes['ng-dirty']).toBeFalsy();
        const input = getInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));
        expect(formField.classes['ng-dirty']).toBeTruthy();
    });

    it('should add ng-valid selector for KbqFormField when form control is valid', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);
        const input = getInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));
        expect(getFormFieldDebugElement(debugElement).classes['ng-valid']).toBeTruthy();
    });

    it('should add ng-invalid selector for KbqFormField initially', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);

        expect(getFormFieldDebugElement(debugElement).classes['ng-invalid']).toBeTruthy();
    });

    it('should add ng-pending selector for KbqFormField when form control is pending', () => {
        const fixture = createComponent(InputFormFieldWithHintAndError);
        const { debugElement, componentInstance } = fixture;
        const formField = getFormFieldDebugElement(debugElement);

        expect(formField.classes['ng-pending']).toBeFalsy();
        componentInstance.control.markAsPending();
        fixture.detectChanges();
        expect(formField.classes['ng-pending']).toBeTruthy();
    });

    it('should add cdk-focused selector for form-field by calling focus() method', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);
        const formField = getFormFieldDebugElement(debugElement);

        expect(formField.classes['cdk-focused']).toBeFalsy();
        getFormFieldDebugElement(debugElement).componentInstance.focus();
        expect(formField.classes['cdk-focused']).toBeTruthy();
    });

    it('should add cdk-focused selector for form-field by calling native input focus() method', () => {
        const { debugElement } = createComponent(InputFormFieldWithHintAndError);
        const formField = getFormFieldDebugElement(debugElement);

        expect(formField.classes['cdk-focused']).toBeFalsy();
        getInputNativeElement(debugElement).focus();
        expect(formField.classes['cdk-focused']).toBeTruthy();
    });

    it('should add kbq-disabled selector for KbqFormField when form control is disabled', () => {
        const fixture = createComponent(InputFormFieldWithHintAndError);
        const { debugElement, componentInstance } = fixture;
        const formField = getFormFieldDebugElement(debugElement);

        expect(formField.classes['kbq-disabled']).toBeFalsy();
        componentInstance.control.disable();
        fixture.detectChanges();
        expect(formField.classes['kbq-disabled']).toBeTruthy();
    });

    it('should add kbq-form-field_invalid selector for KbqFormField initially', () => {
        const { debugElement } = createComponent(InputFormFieldWithCustomErrorStateMatcher);

        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_invalid']).toBeTruthy();
    });

    it('should add kbq-form-field_invalid selector for KbqFormField initially', () => {
        const { debugElement } = createComponent(InputFormFieldWithCustomErrorStateMatcher);

        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_invalid']).toBeTruthy();
    });

    it('should display KbqError initially', async () => {
        const { debugElement } = createComponent(InputFormFieldWithCustomErrorStateMatcher);

        expect(getErrorDebugElement(debugElement)).toBeTruthy();
    });

    it('should add kbq-form-field_invalid selector for KbqFormField on form submission', () => {
        const fixture = createComponent(InputFormFieldWithCustomErrorStateMatcher);
        const { debugElement, componentInstance } = fixture;

        componentInstance.errorStateMatcher = new ShowOnFormSubmitErrorStateMatcher();
        fixture.detectChanges();
        const formField = getFormFieldDebugElement(debugElement);

        expect(formField.classes['kbq-form-field_invalid']).toBeFalsy();
        getSubmitButtonNativeElement(debugElement).click();
        expect(formField.classes['kbq-form-field_invalid']).toBeTruthy();
    });

    it('should display KbqError on form submission', () => {
        const fixture = createComponent(InputFormFieldWithCustomErrorStateMatcher);
        const { debugElement, componentInstance } = fixture;

        componentInstance.errorStateMatcher = new ShowOnFormSubmitErrorStateMatcher();
        fixture.detectChanges();
        expect(getErrorDebugElement(debugElement)).toBeFalsy();
        getSubmitButtonNativeElement(debugElement).click();
        expect(getErrorDebugElement(debugElement)).toBeTruthy();
    });

    it('should display KbqLabel', () => {
        const { debugElement } = createComponent(InputFormFieldWithLabel);

        expect(getLabelDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should add attribute.for for label', () => {
        const { debugElement, componentInstance } = createComponent(InputFormFieldWithLabel);

        expect(getLabelNativeElement(debugElement).getAttribute('for')).toBe(componentInstance.id);
    });

    it('should focus input by click on label', () => {
        const { debugElement } = createComponent(InputFormFieldWithLabel);
        const input = getInputNativeElement(debugElement);

        expect(document.activeElement).not.toBe(input);
        getLabelDebugElement(debugElement).nativeElement.click();
        expect(document.activeElement).toBe(input);
    });

    it('should add kbq-form-field_no-borders selector for KbqFormField by attribute', () => {
        const { debugElement } = createComponent(InputFormFieldWithBorderCustomization);

        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_no-borders']).toBeTruthy();
    });

    it('should remove kbq-form-field_no-borders selector for KbqFormField by attribute', () => {
        const fixture = createComponent(InputFormFieldWithBorderCustomization);
        const { debugElement, componentInstance } = fixture;

        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_no-borders']).toBeTruthy();
        componentInstance.noBorders = false;
        fixture.detectChanges();
        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_no-borders']).toBeFalsy();
    });

    it('should add kbq-form-field_no-borders selector for KbqFormField by KBQ_FORM_FIELD_DEFAULT_OPTIONS', () => {
        const { debugElement } = createComponent(InputFormFieldWithLabel, [
            kbqFormFieldDefaultOptionsProvider({ noBorders: true })
        ]);

        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_no-borders']).toBeTruthy();
    });

    it('should hide KbqPasswordToggle initially', () => {
        const { debugElement } = createComponent(PasswordFormField);

        expect(getPasswordToggleDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should display KbqPasswordToggle by input', () => {
        const { debugElement } = createComponent(PasswordFormField);
        const input = getPasswordInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));
        expect(getPasswordToggleDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should display password by click on KbqPasswordToggle', () => {
        const { debugElement } = createComponent(PasswordFormField);
        const input = getPasswordInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));
        expect(input.type).toBe('password');
        getPasswordToggleDebugElement(debugElement).nativeElement.click();
        expect(input.type).toBe('text');
    });

    it('should change KbqPasswordToggle icon by click', () => {
        const { debugElement } = createComponent(PasswordFormField);
        const input = getPasswordInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));
        const passwordToggle = getPasswordToggleDebugElement(debugElement);

        passwordToggle.nativeElement.click();
        expect(passwordToggle).toMatchSnapshot();
    });

    it('should display KbqReactivePasswordHint initially', () => {
        const { debugElement } = createComponent(PasswordFormField);

        expect(getReactivePasswordHintDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should display KbqReactivePasswordHint error for invalid password', () => {
        const { debugElement } = createComponent(PasswordFormField);
        const input = getPasswordInputNativeElement(debugElement);

        input.value = 'koobiq';
        input.dispatchEvent(new Event('input'));

        expect(getReactivePasswordHintDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should display KbqReactivePasswordHint success for valid password', () => {
        const { debugElement } = createComponent(PasswordFormField);
        const input = getPasswordInputNativeElement(debugElement);

        input.value = 'koobiq-is-awesome';
        input.dispatchEvent(new Event('input'));

        expect(getReactivePasswordHintDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should apply legacy KbqValidateDirective by default', () => {
        const { debugElement } = createComponent(InputFormFieldWithLegacyValidationDirective);

        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_has-validate-directive']).toBeTruthy();
        expect(getInputDebugElement(debugElement).classes['kbq-control_has-validate-directive']).toBeTruthy();
    });

    it('should disable legacy KbqValidateDirective by kbqDisableLegacyValidationDirective() provider', () => {
        const { debugElement } = createComponent(InputFormFieldWithLegacyValidationDirective, [
            kbqDisableLegacyValidationDirectiveProvider()
        ]);

        expect(getFormFieldDebugElement(debugElement).classes['kbq-form-field_has-validate-directive']).toBeFalsy();
        expect(getInputDebugElement(debugElement).classes['kbq-control_has-validate-directive']).toBeFalsy();
    });
});
