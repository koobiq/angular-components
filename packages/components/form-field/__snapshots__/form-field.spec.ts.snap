// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbqFormField should change KbqPasswordToggle icon by click 1`] = `
DebugElement {
  "nativeNode": <kbq-password-toggle
    aria-hidden="false"
    class="kbq-password-toggle"
    style="visibility: visible;"
  >
    <i
      class="kbq kbq-icon-button kbq-icon kbq-contrast-fade kbq-eye-slash_16"
      color="contrast-fade"
      kbq-icon-button=""
      tabindex="0"
    />
  </kbq-password-toggle>,
}
`;

exports[`KbqFormField should display KbqCleaner 1`] = `
DebugElement {
  "nativeNode": <kbq-cleaner
    class="kbq-cleaner"
  >
    <i
      class="kbq kbq-icon-button kbq-icon kbq-contrast-fade kbq-xmark-circle_16"
      color="contrast-fade"
      kbq-icon-button="kbq-xmark-circle_16"
      tabindex="0"
    />
  </kbq-cleaner>,
}
`;

exports[`KbqFormField should display KbqError 1`] = `
DebugElement {
  "nativeNode": <kbq-error
    class="kbq-error kbq-hint"
    id="test-error-id"
  >
    <span
      class="kbq-hint__text"
    >
      Error
    </span>
  </kbq-error>,
}
`;

exports[`KbqFormField should display KbqHint 1`] = `
DebugElement {
  "nativeNode": <kbq-hint
    class="kbq-hint kbq-empty"
    id="test-hint-id"
  >
    <span
      class="kbq-hint__text"
    >
      Hint
    </span>
  </kbq-hint>,
}
`;

exports[`KbqFormField should display KbqLabel 1`] = `
DebugElement {
  "nativeNode": <kbq-label
    class="kbq-label"
  >
    Label
  </kbq-label>,
}
`;

exports[`KbqFormField should display KbqPasswordToggle by input 1`] = `
DebugElement {
  "nativeNode": <kbq-password-toggle
    aria-hidden="false"
    class="kbq-password-toggle"
    style="visibility: visible;"
  >
    <i
      class="kbq kbq-icon-button kbq-icon kbq-contrast-fade kbq-eye_16"
      color="contrast-fade"
      kbq-icon-button=""
      tabindex="0"
    />
  </kbq-password-toggle>,
}
`;

exports[`KbqFormField should display KbqPrefix 1`] = `
DebugElement {
  "nativeNode": <span
    class="kbq-prefix"
    kbqprefix=""
  >
    Prefix
  </span>,
}
`;

exports[`KbqFormField should display KbqReactivePasswordHint error for invalid password 1`] = `
DebugElement {
  "nativeNode": <kbq-reactive-password-hint
    class="kbq-reactive-password-hint kbq-hint kbq-contrast-fade kbq-hint_fill-text-off"
    id="test-reactive-password-hint-id"
  >
    <i
      class="kbq kbq-icon kbq-empty kbq-xmark-s_16"
      kbq-icon=""
    />
    <span
      class="kbq-hint__text"
    >
       Min length 
       (6/8) 
      
    </span>
  </kbq-reactive-password-hint>,
}
`;

exports[`KbqFormField should display KbqReactivePasswordHint initially 1`] = `
DebugElement {
  "nativeNode": <kbq-reactive-password-hint
    class="kbq-reactive-password-hint kbq-hint kbq-contrast-fade kbq-hint_fill-text-off"
    id="test-reactive-password-hint-id"
  >
    <i
      class="kbq kbq-icon kbq-empty kbq-xmark-s_16"
      kbq-icon=""
    />
    <span
      class="kbq-hint__text"
    >
       Min length 
       (0/8) 
      
    </span>
  </kbq-reactive-password-hint>,
}
`;

exports[`KbqFormField should display KbqReactivePasswordHint success for valid password 1`] = `
DebugElement {
  "nativeNode": <kbq-reactive-password-hint
    class="kbq-reactive-password-hint kbq-hint kbq-hint_fill-text-off kbq-success"
    id="test-reactive-password-hint-id"
  >
    <i
      class="kbq kbq-icon kbq-empty kbq-check-s_16"
      kbq-icon=""
    />
    <span
      class="kbq-hint__text"
    >
       Min length 
      
    </span>
  </kbq-reactive-password-hint>,
}
`;

exports[`KbqFormField should display KbqSuffix 1`] = `
DebugElement {
  "nativeNode": <span
    class="kbq-suffix"
    kbqsuffix=""
  >
    Suffix
  </span>,
}
`;

exports[`KbqFormField should hide KbqPasswordToggle initially 1`] = `
DebugElement {
  "nativeNode": <kbq-password-toggle
    aria-hidden="true"
    class="kbq-password-toggle cdk-visually-hidden"
    style="visibility: hidden;"
  >
    <i
      class="kbq kbq-icon-button kbq-icon kbq-contrast-fade kbq-eye_16"
      color="contrast-fade"
      kbq-icon-button=""
      tabindex="0"
    />
  </kbq-password-toggle>,
}
`;
